rfcs/2020-04-23-2445-log-schema.md

```diff
@@ -0,0 +1,239 @@
+   a [`source` and `source_type` field][splunk_source_fields] that describes
+   where the data came from (ex: `source` = `TCP:0.0.0.0:4000`).
+4. To solve for this we [introduced a new global `log_schema.*`][pr_1769] set
+   of options that contained only 3 keys: `log_schema.message_key`,
+   `log_schema.timestamp_key`, `log_schema.host_key`.
+
+In addition to the above, other issues have popped up that would be solved with
+the proper log schema:
+
+1. Internal metadata ([#2415])
+2. Adopting existing schemas (]#2423])
+
+## Proposal
+
+1. Move all component level keys to the global `log_schema` set of options.
+2. Remove all component level `*_key` options in favor of the new `log_schema.*`
+   options.
+3. Define a namespace for metadata, such as `log_schema.meta_key`. This is
+   intended for internal Vector metadata that does not describe the log
+   directly.
+4. If the field does not directly describe the log, or it's origine, then it
+   should go under the new `log_schema.meta_key`.
+5. Setting any of these fields to `""` disables the field.
+6. Offer a "legacy" setting that sets all of the fields below to their current
+   values. This will ensure that users upgrading do not have to change their
+   field names.
+
+Some rules to go by:
+
+1. The _option_ names should be _specific_, we should avoid abstractions where
+   possible. Options names != attribute names. For example, instead of
+   `log_schema.cloud.instance_id_key` it should be
+   `log_schema.aws_ec2.instance_id_key`. If the user want's to generalize all of
+   this in a `"cloud"` namespace they can simply set the value of
+   `log_schema.aws_ec2_key` to `"cloud"`.
+2. We should favor namespacing this data since it'll help to avoid breaking
+   changes as this matures. For example, instead of `socket` we should have
+   a `socket.address` key.
+3. We should try to avoid bikeshedding event names as much as possible. The
+   options should correlate 1 to 1 with every field we touch in Vector. We can
+   default the names to whatever we want.
+
+## Prior Art
+
+1. [Elastic Common Schema][ecs]
+2. [Timber Schema][timber_schema]
+3. [OpenTelemtry log data model proposal][ot_log_data_model]
+
+## Sales Pitch
+
+1. The Vector log schema controls will be discoverable and editable in one
+   place.
+2. It creates an explicit contract across Vector.
+3. It increases Vector's adoption rate. The ability to easily adopt an existing
+   log schema is critical to acquiring larger users with legacy systems. Log
+   schemas are arguably the biggest source of lock-in when coming from other
+   tools.
+4. We can be flexible and adopt new standards as they emerge. For example,
+   OpenTelemetry is [working on a schema standard][ot_log_data_model] that I
+   would like Vector to _optionally_ adopt when ready.
+
+## Drawbacks
+
+1. It's more complicated.
+2. I'm unsure if this will affect performance negatively. Adding more fields
```

Comment:
It shouldn't affect performance any more than the alternatives. And with the ability to turnoff the fields, users could use it to trade extra fields for extra performance. Some of the presets could also, by default, turn off some of them. Example: `slim` preset could have only the most important fields set, and others disabled.